prov/gni/src/gnix_datagram.c

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright (c) 2015 Cray Inc.  All rights reserved.
+ * Copyright (c) 2015 Los Alamos National Security, LLC. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <assert.h>
+
+#include <rdma/fabric.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_prov.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_rma.h>
+#include <rdma/fi_errno.h>
+
+#include "gnix.h"
+#include "gnix_util.h"
+
+int gnix_dgram_alloc(struct gnix_dgram_hndl *hndl, enum gnix_dgram_type type,
+			struct gnix_datagram **d_ptr)
+{
+	int ret = -FI_ENOMEM;
+	struct gnix_datagram *d = NULL;
+	struct list_head *the_free_list;
+	struct list_head *the_active_list;
+
+	if (type == GNIX_DGRAM_WC) {
+		the_free_list = &hndl->wc_datagram_free_list;
+		the_active_list = &hndl->wc_datagram_active_list;
+	} else {
+		the_free_list = &hndl->datagram_free_list;
+		the_active_list = &hndl->datagram_active_list;
+	}
+
+	if (!list_empty(the_free_list)) {
+		d = list_top(the_free_list, struct gnix_datagram, list);
+		if (d != NULL) {
+			gnix_list_del_init(&d->list);
+			list_add(the_active_list, &d->list);
+			d->type = type;
+			ret = FI_SUCCESS;
+		}
+	}
+
```

Comment:
No, we only want to have a finite number of datagrams that the upper layer of gnix can use.  The reason for this is that we don't want to flood the kgni "session" state engine with too many requests.  The provider will be allocating a datagram when its trying to establish a smsg based "connection" with a peer.  If the datagram alloc fails, the gni provider will need to put this connection request on to a pending connection queue and wait till one of the previous connection requests completes, thereby freeing a datagram.

Yes this is an example of stuff I need to put into a design doc.